### PR TITLE
Use new NegotiateAuthentication API instead of reflection of internal API

### LINKED ISSUE #42323: Update negotate auth to replace reflection with new public API
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

The current reflection approach is brittle and not trim-friendly: https://github.com/dotnet/runtime/issues/29270#issuecomment-1126881385

### Describe the solution you'd like

1. Replace with new API: https://github.com/dotnet/runtime/pull/70720
2. Remove trim warnings from negotiate auth

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by filipnavara**
# Use new NegotiateAuthentication API instead of reflection of internal API

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

Replaces the reflection over internal `NTAuthentication` API in System.Net.Security with the new public API.

Notes:
- The lifetime of `NegotiateAuthentication.RemoteIdentity` is bound to the lifetime of `NegotiateAuthentication` object. That's different from `NTAuthentication` where the caller was responsible for disposing it (if it is `IDisposable`).

Fixes #42323

### CODE REVIEW

File: src/Security/Authentication/Negotiate/src/NegotiateHandler.cs [hunk 1]
```diff
@@ -319,7 +319,6 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
         if (OperatingSystem.IsWindows() && identity is WindowsIdentity winIdentity)
         {
             user = new WindowsPrincipal(winIdentity);
-            Response.RegisterForDispose(winIdentity);
```
Comment (Tratcher) [inline src/Security/Authentication/Negotiate/src/NegotiateHandler.cs:322]: See the comments on lines 315-6. If NegotiateState returns us the same instance each time, we need to make a copy of it for use on this request.

File: src/Security/Authentication/Negotiate/src/Internal/NegotiateState.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    private static bool IsClientError(NegotiateAuthenticationStatusCode error)
+    {
+        return error == NegotiateAuthenticationStatusCode.InvalidToken ||
+            error == NegotiateAuthenticationStatusCode.QopNotSupported ||
+            error == NegotiateAuthenticationStatusCode.UnknownCredentials ||
+            error == NegotiateAuthenticationStatusCode.MessageAltered ||
+            error == NegotiateAuthenticationStatusCode.OutOfSequence ||
+            error == NegotiateAuthenticationStatusCode.InvalidCredentials;
+    }
```
Comment (Tratcher) [inline src/Security/Authentication/Negotiate/src/Internal/NegotiateState.cs:85]: There are a lot more error codes in the old logic. Where they consolidated? Or would some never really happen?

Comment (filipnavara) [reply src/Security/Authentication/Negotiate/src/Internal/NegotiateState.cs:85]: They were consolidated (lot of the status codes were from unrelated Schannel APIs for TLS connections).

File: src/Security/Authentication/Negotiate/src/NegotiateHandler.cs (hunk 1 above)
Comment (filipnavara) [reply src/Security/Authentication/Negotiate/src/NegotiateHandler.cs:322]: I understood that as the `WindowsPrincipal` object being mutable and that's still created per request. However, if the identity itself is modified too then a copy would be in order.

Comment (Tratcher) [reply src/Security/Authentication/Negotiate/src/NegotiateHandler.cs:322]: Yes, the identity is mutable and needs to be copied.

File: src/Security/Authentication/Negotiate/src/Internal/NegotiateState.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    private readonly NegotiateAuthentication _instance;
+
+    public NegotiateState()
+    {
+        _instance = new NegotiateAuthentication(_serverOptions);
+    }
+
+    public string? GetOutgoingBlob(string incomingBlob, out BlobErrorType status, out Exception? error)
+    {
+        string? outgoingBlob = _instance.GetOutgoingBlob(incomingBlob, out NegotiateAuthenticationStatusCode authStatus);
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/Security/Authentication/Negotiate/src/Internal/NegotiateState.cs]: Fixed.

**Review by Tratcher — APPROVED** [review]

**Review by JamesNK — CHANGES_REQUESTED** [review]
There are some places where RequiresUnreferencedCode is still present. See https://cs.github.com/dotnet/aspnetcore?q=Negotiate+authentication+uses+types+that+cannot+be+statically+analyzed

For example, [NegotateHandler](https://github.com/dotnet/aspnetcore/blob/dca507ad67c057c2332c67a612f9538520ce29a4/src/Security/Authentication/Negotiate/src/NegotiateHandler.cs#L22)

**Comment by JamesNK** [comment]
I think NegotateHandler and NegotateExtensions have the remaining attributes to remove: https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authentication/Negotiate/src/NegotiateExtensions.cs

**Comment by filipnavara** [comment]
@JamesNK hopefully addressed now.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**